pkg/network/tracer/process_cache.go

```diff
@@ -0,0 +1,300 @@
+	if p == nil {
+		pc.stats.eventsSkipped.Add(1)
+		return
+	}
+
+	select {
+	case pc.in <- p:
+	default:
+		// dropped
+		pc.stats.eventsDropped.Add(1)
+	}
+}
+
+func (pc *processCache) processEvent(entry *smodel.ProcessCacheEntry) *process {
+	var envs map[string]string
+	if entry.EnvsEntry != nil {
+		for _, v := range entry.EnvsEntry.Values {
+			kv := strings.SplitN(v, "=", 2)
+			k := kv[0]
+
+			if len(pc.filteredEnvs) > 0 {
+				if _, found := pc.filteredEnvs[k]; !found {
+					continue
+				}
+			}
+
+			v = ""
+			if len(kv) > 1 {
+				v = kv[1]
+			}
+
+			if envs == nil {
+				envs = make(map[string]string)
+			}
+			envs[k] = v
+			if len(pc.filteredEnvs) > 0 && len(pc.filteredEnvs) == len(envs) {
+				break
+			}
+		}
+	}
+
+	if len(envs) == 0 && len(pc.filteredEnvs) > 0 && entry.ContainerID == "" {
+		return nil
+	}
+
+	return &process{
+		Pid:         entry.Pid,
+		Envs:        envs,
+		ContainerID: entry.ContainerID,
+		StartTime:   entry.ExecTime.UnixNano(),
+	}
+}
+
+func (pc *processCache) Stop() {
+	if pc == nil {
+		return
+	}
+
+	pc.stop.Do(func() { close(pc.stopped) })
+}
+
+func (pc *processCache) add(p *process) {
+	if pc == nil {
+		return
+	}
+
+	pc.Lock()
+	defer pc.Unlock()
+
+	evicted := pc.cache.Add(processCacheKey{pid: p.Pid, startTime: p.StartTime}, p)
+	pl, _ := pc.cacheByPid[p.Pid]
+	pc.cacheByPid[p.Pid] = pl.update(p)
+
+	if evicted {
+		pc.stats.cacheEvicts.Add(1)
+	}
+}
+
+func (pc *processCache) GetStats() map[string]interface{} {
+	if pc == nil {
+		return map[string]interface{}{}
+	}
+
+	pc.Lock()
+	defer pc.Unlock()
+
+	pc.stats.cacheLength.Store(uint64(pc.cache.Len()))
+	return atomicstats.Report(&pc.stats)
+}
+
+func (pc *processCache) Get(pid uint32, ts int64) (*process, bool) {
+	if pc == nil {
+		return nil, false
+	}
+
+	pc.Lock()
+	defer pc.Unlock()
+
+	pl, _ := pc.cacheByPid[pid]
+	if closest := pl.closest(ts); closest != nil {
+		pc.cache.Get(processCacheKey{pid: closest.Pid, startTime: closest.StartTime})
+		return closest, true
+	}
+
+	return nil, false
+}
+
+func (pc *processCache) Dump() (interface{}, error) {
+	res := map[uint32]interface{}{}
+	if pc == nil {
+		return res, nil
+	}
+
+	pc.Lock()
+	defer pc.Unlock()
+
+	for pid, pl := range pc.cacheByPid {
+		res[pid] = pl
+	}
+
+	return res, nil
+}
+
+func (pl processList) update(p *process) processList {
+	for i := range pl {
+		if pl[i].Pid == p.Pid && pl[i].StartTime == p.StartTime {
+			pl[i] = p
+			return pl
+		}
+	}
+
+	if len(pl) == maxProcessListSize {
+		pl = pl[1:]
+	}
+
+	return append(pl, p)
+}
+
+func (pl processList) remove(p *process) processList {
+	for i := range pl {
+		if pl[i] == p {
+			return append(pl[:i], pl[i+1:]...)
+		}
+	}
+
+	return pl
+}
+
+func (pl processList) closest(ts int64) *process {
+	var closest *process
+	for i := range pl {
+		if ts < pl[i].StartTime {
+			continue
+		}
+
+		if closest == nil {
+			closest = pl[i]
+			continue
+		}
+
+		if pl[i].StartTime > closest.StartTime {
```

Comment:
This seems to be choosing the latest process, rather than the one closest to the timestamp?

Reply:
Fixed